lib/rust/profiling/src/frame_stats.rs

```diff
@@ -0,0 +1,236 @@
+//! This module helps to aggregate various per-frame performance statistics, collected over labeled
+//! time intervals.
+#![warn(missing_debug_implementations)]
+#![warn(missing_docs)]
+#![warn(trivial_casts)]
+#![warn(trivial_numeric_casts)]
+#![warn(unsafe_code)]
+#![warn(unused_import_braces)]
+
+
+use enso_prelude::*;
+
+use enso_logger::DefaultWarningLogger as Logger;
+use enso_logger::*;
+use serde::Deserialize;
+use serde::Serialize;
+
+
+
+// ====================
+// === IntervalsMap ===
+// ====================
+
+type IntervalsMap = HashMap<String, Bundle>;
+
+thread_local! {
+    static ACTIVE_INTERVALS: RefCell<IntervalsMap> = RefCell::new(IntervalsMap::new());
+}
+
+/// Starts a new named time interval, during which frame statistics will be collected.
+pub fn start_interval(label: &str) {
+    let logger = Logger::new("Profiling_Stats");
+    ACTIVE_INTERVALS.with(|intervals| {
+        let found = intervals.borrow_mut().insert(label.to_string(), Bundle::default());
+        if found.is_some() {
+            warning!(logger, "Trying to collect profiling stats for a process with same label as already existing one - values will be skewed for: {label:?}");
+        }
+    });
+}
+
+/// Finishes collecting frame statistics for a specific named interval. Returns aggregate data
+/// collected since the start of the the interval.
+pub fn end_interval(label: &str) -> Option<Bundle> {
```

Comment:
I was thinking about RAII-like API with the following signatures: `start_interval(label: &str) -> IntervalGuard`, `IntervalGuard::end(&self) -> Option<Bundle>`. But I guess it doesn't work well with the places where we want to add this?

Reply:
I tried following this idea, but this seems notably incompatible with the current `mark_end_interval()` func's signature in the base branch. I'm not sure whether that func is going to stay there or not in the end, given that the base branch of this PR seems to be still in flux and under discussion. For now, I think I'll polish a few minor things on this PR (like a few more tests), and call a blocker on this task/PR, on basis of the base branch being in flux. From looking at the comments I'm getting from you and @farmaazon, there is definitely some potential space for improvement in this PR, but none of them seem to be calling out issues with the core architecture of the implementation (i.e. hashmap etc.), so in the normal process, I would probably expect to be addressing them once this PR is "promoted" to a non-draft. Which I'm not convinced I should be doing yet if the base branch has potential for even some major API changes.

I would be happy to revisit this idea once the base branch stabilizes.

Reply:
In later iterations, I managed to implement a RAII-like API. Thanks for the suggestion!